packages/wrangler/src/dev/start-server.ts

```diff
@@ -131,26 +143,75 @@ export async function startDevServer(
 
 				return { accountId, apiToken: requireApiToken() };
 			},
+			persist: props.localPersistencePath ?? undefined,
+			testScheduled: props.testScheduled,
+			registry: workerDefinitions,
 		},
 		build: {
-			// format: props.entry.format,
+			bundle: !props.noBundle,
+			define: props.define,
+			jsxFactory: props.jsxFactory,
+			jsxFragment: props.jsxFragment,
+			tsconfig: props.tsconfig,
+			minify: props.minify,
+			processEntrypoint: props.processEntrypoint,
+			additionalModules: props.additionalModules,
 			moduleRoot: props.entry.moduleRoot,
-			nodejsCompatMode: null,
+			moduleRules: props.rules,
+			nodejsCompatMode: props.nodejsCompatMode,
 		},
 	};
 
-	// temp: fake these events by calling the handler directly
-	if (!props.experimentalDevEnv) {
-		devEnv.proxy.onConfigUpdate({
-			type: "configUpdate",
-			config: fakeResolvedInput(startDevWorkerOptions),
-		});
-		devEnv.proxy.onBundleStart({
-			type: "bundleStart",
-			config: fakeResolvedInput(startDevWorkerOptions),
+	if (props.experimentalDevEnv) {
+		// to comply with the current contract of this function, call props.onReady on reloadComplete
+		devEnv.runtimes.forEach((runtime) => {
+			runtime.on("reloadComplete", async (ev) => {
+				const { proxyWorker } = await devEnv.proxy.ready.promise;
+				const url = await proxyWorker.ready;
+
+				props.onReady?.(url.hostname, parseInt(url.port), ev.proxyData);
+			});
 		});
+
+		await devEnv.config.set(startDevWorkerOptions);
+
+		await Promise.all([
+			// adhere to unstable_dev contract:
+			//   - only resolve when UserWorker is ready
+			//   - reject if UserWorker fails to start
+			Promise.race(
+				devEnv.runtimes.flatMap((runtime) => [
+					once(runtime, "reloadComplete"),
+					once(runtime, "error").then((err) => Promise.reject(err)),
+				])
+			),
+			// adhere to unstable_dev contract:
+			//   - only resolve when _perceived_ UserWorker is ready
+			//   - throw if _perceived_ UserWorker fails to start
+			// to the eyeball, the ProxyWorker is the _perceived_ UserWorker
+			Promise.race([
+				devEnv.proxy.ready.promise,
+				once(devEnv.proxy, "error").then((err) => Promise.reject(err)),
+			]),
+		]);
+
+		return {
```

Comment:
Made this function clearer by consolidating into one if-block and returning early when props.experimentalDevEnv is set. The rest of the function is for when it is not set